### PR TITLE
Refactor DFE to use pyquil measure_observables

### PR DESCRIPTION
**Pull request by mpharrigan**
In the spirit of #4 and #17, this re-writes DFE to use the shared set of datastructures exposed in pyquil's operator_estimation.py.

As a bonus: I've implemented monte-carlo DFE, and for exhaustive dfe I've made it exhaustive over the input states (+1 and -1 eigenstates for each of X, Y, Z instead of always preparing the +1 eigenstate). This addresses the same issue as #29.

One large caveat is that the `acquire` function does readout symetrization and calibration which pyquil's `measure_observables` has not learned how to do yet. For full feature parity, the symmetrization and calibration code should be ported to those routines. As a benefit, readout calibration will automatically become available and be shared among dfe, state, and process tomography as well as any e.g. VQE-style algorithms that use `measure_observables`

### CODE REVIEW

File: forest_benchmarking/dfe.py [hunk 1]
```diff
@@ -51,66 +41,128 @@ def calibrate_readout_imperfections(pauli: PauliTerm, quantum_machine: QuantumCo
...
+        if oneq_st.label == 'X':
+            term *= sX(oneq_st.qubit)
+        elif oneq_st.label == 'Y':
+            term *= sY(oneq_st.qubit)
+        elif oneq_st.label == 'Z':
+            term *= sZ(oneq_st.qubit)
+        else:
+            raise ValueError(f"Can't convert state {state} to a PauliTerm")
+
+        if oneq_st.index == 1:
```
Comment (msohaibalam) [inline forest_benchmarking/dfe.py]: This is probably more a statement about the structure of `_OneQState` (in pyquil/operator_estimation) than the code here, but using `index` to denote the eigenstate can potentially be quite confusing: almost everywhere else in our code-base, we use index to denote the qubit on which the operation is occurring.

Comment (mpharrigan) [reply forest_benchmarking/dfe.py]: we should move towards using "qubit" to denote qubits. Most of the gates use this convention

File: forest_benchmarking/dfe.py [hunk 2]
```diff
@@ -1,36 +1,26 @@
...
-  + make sure pre-measurement rotations happen simultaneously (i.e., don't get schedules
-    "into" program being characterized)
-  + write various tests cases for automation
-     + CZ fidelity with neighbours in ground state
-     + CZ fidelity with neighbours in excited state
-     + sequence of graph states on subset of some graph
-        + linear chain
-        + 4.8.8
-        + etc
-"""
```
[the remark this answers is not in the document]
Comment (mpharrigan) [reply forest_benchmarking/dfe.py]: Yes, they should be issues

[the remark this answers is not in the document]
Comment (joshcombes) [reply forest_benchmarking/dfe.py]: done

File: forest_benchmarking/dfe.py [hunk 3]
```diff
@@ -140,7 +192,7 @@ class DFEdata:
     count: List[int]
     """number of shots used to calculate the `expectation`"""
 
-def acquire_dfe_data(experiment: DFEexperiment, quantum_machine: QuantumComputer, var: float = 0.01) -> Tuple[DFEdata, DFEdata]:
+def acquire_dfe_data(experiment, quantum_machine: QuantumComputer, var: float = 0.01) -> Tuple[DFEdata, DFEdata]:
```
Comment (kylegulshen) [inline forest_benchmarking/dfe.py]: Don't we need a shim for any of the additions to be used here (Why remove the DFEexperiment tag?)?

File: forest_benchmarking/dfe.py (hunk 1 above)
Comment (msohaibalam) [reply forest_benchmarking/dfe.py]: Sounds good to me.

**Comment by msohaibalam** [comment]
I'm merging this branch to master. Any other issues that remain unresolved will be worked on in later PR(s).

**Pull request merged**